src/java/org/apache/cassandra/index/sai/disk/v1/postings/PostingsWriter.java

```diff
@@ -0,0 +1,314 @@
+ * to encode most data as a packed block. Take a term with 259 row IDs as an example, the first 256 IDs are encoded
+ * as two packed blocks, while the remaining 3 are encoded as one VLong block.
+ * </p>
+ * <p>
+ * Each posting list ends with a meta section and a skip table, that are written right after all postings blocks. Skip
+ * interval is the same as block size, and each skip entry points to the end of each block.  Skip table consist of
+ * block offsets and last values of each block, compressed as two FoR blocks.
+ * </p>
+ *
+ * Visual representation of the disk format:
+ * <pre>
+ *
+ * +========+========================+=====+==============+===============+============+=====+========================+========+
+ * | HEADER | POSTINGS LIST (TERM 1)                                                   | ... | POSTINGS LIST (TERM N) | FOOTER |
+ * +========+========================+=====+==============+===============+============+=====+========================+========+
+ *          | FOR BLOCK (1)          | ... | FOR BLOCK (N)| BLOCK SUMMARY              |
+ *          +------------------------+-----+--------------+---------------+------------+
+ *                                                        | BLOCK SIZE    |            |
+ *                                                        | LIST SIZE     | SKIP TABLE |
+ *                                                        +---------------+------------+
+ *                                                                        | BLOCKS POS.|
+ *                                                                        | MAX VALUES |
+ *                                                                        +------------+
+ *
+ *  </pre>
+ */
+@NotThreadSafe
+public class PostingsWriter implements Closeable
+{
+    private static final String POSTINGS_MUST_BE_SORTED_ERROR_MSG = "Postings must be sorted ascending, got [%s] after [%s]";
+
+    private final IndexOutput dataOutput;
+    private final int blockSize;
+    private final long[] deltaBuffer;
+    private final LongArrayList blockOffsets = new LongArrayList();
+    private final LongArrayList blockMaxIDs = new LongArrayList();
+    private final RAMIndexOutput inMemoryOutput = new RAMIndexOutput("blockOffsets");
+
+    private final long startOffset;
+
+    private int bufferUpto;
+    private long lastSegmentRowId;
+    private long maxDelta;
+    private long totalPostings;
+
+    public PostingsWriter(IndexDescriptor indexDescriptor, IndexContext indexContext, boolean segmented) throws IOException
+    {
+        this(indexDescriptor, indexContext, BLOCK_SIZE, segmented);
+    }
+
+    public PostingsWriter(IndexOutput dataOutput) throws IOException
+    {
+        this(dataOutput, BLOCK_SIZE);
+    }
+
+    @VisibleForTesting
+    PostingsWriter(IndexDescriptor indexDescriptor, IndexContext indexContext, int blockSize, boolean segmented) throws IOException
```

Comment:
Nit: We could make `blockSize` the last argument, so this constructor is identical to the previous one but with one additional argument at the end.